BedrockServer.cpp

```diff
@@ -1862,3 +1914,87 @@ void BedrockServer::_acceptSockets() {
     }
 }
 
+void BedrockServer::_loadSecureData(const map<BedrockPlugin*, list<string>>& pluginSecureDataMap, const SData& args) {
+    // Do we need to load secure data? Skip loading if we don't. In this case, plugins
+    // should manage loading debug secure data inside of the plugin.
+    if (!args.isSet("-loadSecureData")) {
+        SDEBUG("Skipping loading secure data.");
+        return;
+    }
+
+    // If a scecure data host wasn't specified, use the default.
+    string keyHost;
+    if (!args.isSet("-secureDataHost")) {
+        keyHost = "localhost:4001";
+    } else {
+        keyHost = args["-secureDataHost"];
+    }
+
+    // Loop through the plugins that require secure data, if the file name
+    // is prefixed with socket:, load the data from a socket instead.
+    for (auto plugin : pluginSecureDataMap) {
+        const string& pluginName = SToLower(plugin.first->getName());
+
+        for (auto file : plugin.second) {
+            const string& fileName = file;
+            SData secureData;
+
+            if (!SStartsWith(fileName, "socket")) {
+                // If a secure data file exists for this plugin, try and open it and read from it.
+                if (SFileExists(fileName)) {
+                    FILE* file = fopen(fileName.c_str(), "r");
+                    if (!file) {
+                        SWARN("Couldn't open file secure data file " << fileName << " for reading. Error: " << errno << ", " << strerror(errno) << ".");
+                        continue;
+                    }
+                    SINFO("Successfully opened secure data file " << fileName << " for reading.");
+
+                    // Read the whole file into a buffer, these should be small text files,
+                    // buf if the file is too large we could die here with bad_alloc.
+                    size_t fileSize = SFileSize(fileName);
+                    char* buf = new char[fileSize];
+                    fread(buf, 1, fileSize, file);
+
+                    // Move the file into an SData object and pass it to the plugin it belongs to.
+                    secureData.deserialize(string(buf));
+                    plugin.first->loadSecureData(secureData);
```

Comment:
Why do we make SData's out of these instead of passing the raw string to the plugin?

I'm thinking of a key that looks like an SSL cert or something.

Reply:
Ah yeah, I think this is a good generalization of this code. All of our secure data is currently in the name:value format so SData makes sense, but I can see where a plugin would just load a single string instead.